pkg/controllers/cisnetworkpolicy/controller.go

```diff
@@ -0,0 +1,160 @@
+	core "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	FlannelPresenceLabel         = "flannel.alpha.coreos.com/public-ip"
+	FlannelHostNetworkPolicyName = "rke2-flannel-host-networking"
+)
+
+func CISNetworkPolicyController(ctx context.Context, sc *server.Context) error {
+	return register(ctx, sc.Core.Core().V1().Node(), sc.K8s)
+}
+
+func register(ctx context.Context,
+	nodes coreclient.NodeController,
+	k8s kubernetes.Interface,
+) error {
+	h := &handler{
+		ctx: ctx,
+		k8s: k8s,
+	}
+	logrus.Debugf("CISNetworkPolicyController: Registering controller hooks")
+	nodes.OnChange(ctx, "cisnetworkpolicy-node", h.handle)
+	nodes.OnRemove(ctx, "cisnetworkpolicy-node", h.handle)
+	return nil
+}
+
+type handler struct {
+	ctx context.Context
+	k8s kubernetes.Interface
+}
+
+func (h *handler) handle(key string, node *core.Node) (*core.Node, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	return h.reconcileFlannelHostNetworkPolicy(key, node)
+}
+
+func (h *handler) reconcileFlannelHostNetworkPolicy(nodeName string, node *core.Node) (*core.Node, error) {
+	var np *netv1.NetworkPolicy
+	var npIR *netv1.NetworkPolicyIngressRule
+
+	npIR, err := h.generateHostNetworkPolicyIngressRule()
+	if err != nil {
+		return nil, err
+	}
+
+	np, err = h.generateHostNetworkingNetworkPolicy(npIR)
+	if err != nil {
+		return nil, err
+	}
+
+	var namespaces = []string{
+		metav1.NamespaceSystem,
+		metav1.NamespaceDefault,
+		metav1.NamespacePublic,
+	}
+
+	for _, namespace := range namespaces {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if npGet, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Get(h.ctx, FlannelHostNetworkPolicyName, metav1.GetOptions{}); err != nil {
+				if errors.IsNotFound(err) {
+					if _, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Create(h.ctx, np, metav1.CreateOptions{}); err != nil {
+						if !errors.IsAlreadyExists(err) {
+							logrus.Errorf("CISNetworkPolicyController: Error creating network policy in namespace %s: %v", namespace, err)
+							return err
+						}
+					}
+				} else {
+					logrus.Errorf("CISNetworkPolicyController: Error getting network policy network policy in namespace %s: %v", namespace, err)
+					return err
+				}
+			} else {
+				npGet = npGet.DeepCopy()
+				npGet.Spec.Ingress[0] = *npIR
+				_, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Update(h.ctx, npGet, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("CISNetworkPolicyController: error working on network policy: %v", err)
+		}
+	}
+	logrus.Debugf("CISNetworkPolicyController: Handled node change")
+	return nil, nil
+}
+
+func (h *handler) generateHostNetworkingNetworkPolicy(networkPolicyIngressRule *netv1.NetworkPolicyIngressRule) (*netv1.NetworkPolicy, error) {
```

Comment:
Does this method return an error to satisfy an interface? If not, we should update it to not return an error that's never created.